### PR TITLE
[Wait for #2568] [ Tensor ] add is_NaN check in Tensor  @open sesame 05/10 14:17

### PR DESCRIPTION
**Pull request by jijoongmoon**
## In this PR

This PR adds the is_NaN function to check if the tensor has a NaN value. This
is for the check of NaN during mixed precision training.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/tensor/blas_interface.cpp [hunk 1]
```diff
@@ -1090,4 +1090,37 @@ void ele_div(const unsigned int N, const float *X, const float *Y, float *Z,
     ele_div_fallback(N, X, Y, Z, alpha, beta, i_stride, o_stride);
 }
 
+bool has_nan(const size_t N, ml::train::TensorDim::DataType d_type,
```
Comment (skykongkong8) [inline nntrainer/tensor/blas_interface.cpp:1093]: did not check for other wait-for PRs yet, but commit in THIS pr looks fine

**Review by skykongkong8 — APPROVED** [review]

File: meson.build [hunk 2]
```diff
@@ -64,9 +64,19 @@ warning_c_flags = [
   '-Wno-error=varargs'
 ]
 
+arch = host_machine.cpu_family()
+
+if get_option('enable-avx')
+   extra_defines += '-DUSE_AVX=1'
+   if get_option('platform') == 'tizen'
+      add_project_arguments(['-mavx2'], language: ['c','cpp'])
```
Comment (djeong20) [inline meson.build:72]: would the Tizen platform always support AVX2 instructions?

File: nntrainer/tensor/blas_avx.cpp [hunk 3]
```diff
@@ -114,4 +115,99 @@ void vcvt_f32_f16(size_t N, const float *input, void *output) {
...
 }
 
+bool hasNaN(const size_t N, const _Float16 *input) {
+  assert(N != 0);
+  assert(input != NULL);
+
+  int temp = 0;
+  size_t idx = 0;
+
+  // 16 single-precision check : ( X != X )
```
Comment (djeong20) [inline nntrainer/tensor/blas_avx.cpp:125]:
```suggestion
  // 16 half-precision check : ( X != X )
```

File: nntrainer/tensor/blas_avx.cpp [hunk 4]
```diff
@@ -114,4 +115,99 @@ void vcvt_f32_f16(size_t N, const float *input, void *output) {
...
+      return true;
+
+    __m256 res1 = _mm256_cmp_ps(vec1, vec1, _CMP_NEQ_UQ);
+    temp = temp | _mm256_movemask_ps(res1);
+
+    if (temp)
+      return true;
+  }
+
+  // 8 single-precision check : ( X != X )
```
Comment (djeong20) [inline nntrainer/tensor/blas_avx.cpp:147]:
```suggestion
  // 8 half-precision check : ( X != X )
```

File: packaging/nntrainer.spec [hunk 5]
```diff
@@ -65,6 +65,13 @@
 %define neon_support -Denable-neon=false
 %endif # arch aarch64
 
+%ifarch x86_64
+%define enable_avx 1
+%define avx_support -Denable-avx=true
+%else
+%define avx_support -Denable-avx=false
+%endif # arch aarch64
```
Comment (djeong20) [inline packaging/nntrainer.spec:73]:
```suggestion
%endif # arch x86_64
```

**Comment by jijoongmoon** [comment]
closed by  #2663

**Pull request closed without merging**